Add operation headers

woodard opened the pull request:
Propose adding a new formatting structure when specifying DWARF operations.

woodard committed:
Start converting operations to a new format

Propose adding a new formatting structure when specifying DWARF
operations.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,261 @@
+>     Y: Z
+>     X: Y
+
+In section 3.3 replace the operation desctions as follows:
```
Fix the typo here.

woodard commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,261 @@
+>     X: X
+>     Y: X
+> 
+>    The DW_OP_dup operation duplicates the entry at the top of the stack.
```
I really thought about adding an explict "the stack parameter is not popped" to these items since I said that it was assumed above but in the end I decided that "the people reading this are computer scientists, they are smart and will understand that dup and the other explicit stack operations don't pop their paramters.

woodard commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,261 @@
+>   stack parameters:
+>     X [*any*]
+>   stack output:
+>     X: X
```
This should be Y X.

woodard committed:
Fix mistakes

As a first draft, this had many typos and selling mistakes. I think that
I fixed all the spelling mistakes and several other mistakes that I
noticed along the way.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Start fixing markdown

Start fixing markdown.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Another try at fixing markdown

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Another attempt to fix markdown

Which flavor will github actually render?

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
How does this look

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Fix markdown

Fix the markdown with a WYSIWYG editor.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,260 @@
+> also receive parameters from the stack and make use of information
+> from its evaluation context.
+>
+> *The structure description of the inputs and each of each operation
```
My reasoning was: The typographical convention that I used to specify how I present the operators is not part of the standard. The information that I present when I actually define the operators is the standard. We could completely change how the information is presented and what DWARF is wouldn't change.

woodard commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,260 @@
+>         - X [*any*]
+>       - stack output:
+>         - Y: X
+>         - X: X
```
That was intentional. If the parameter had a meaningful name then I would have made it:

>       - stack parameters:
>         - X [*any*]: parameter
>       - stack output:
>         - Y:  parameter
>         - X:  parameter

but in this case whatever is in X doesn't have a name other than something generic like "parameter". Instead, I'm kind of using "X" as the parameter name so in this context it indicates both "the symbolic name for top of the stack" and "the generic name for whatever is at the top of the stack".

Because of Scott's plea to not use single letter names, I seriously considered not using X, Y, Z, and T for stack positions and instead giving everything symbolic names but when I looked at it, I realized having the letters there providing an additional clue of where things were on the stack. So I kind of decided that it was better to do what I did than to have:

>       - stack parameters:
>         - parameter [*any*]
>       - stack output:
>         - parameter
>         - parameter

but in cases where there is some meaning like there is in DW_OP_overlay I did:

> 1. `DW_OP_overlay`
>      - stack parameters:
>        - T [location]: base location
>        - Z [location]: overlay location
>        - Y [unsigned int]: base offset
>        - X [unsigned int]: overlay width
>       - stack output:
>         - X [location]: composite_location

That was my reasoning. We can change it if people think having X pull double duty is too confusing.

woodard commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,260 @@
+
+> 1. `DW_OP_lit0`, ..., `DW_OP_lit31`
+>       - stack output:
+>         - X [unsigned integer in the range 0-31 inclusive]
```
Is that actually part of the standard or is that an implementation detail? 
Of course that makes sense but could a conforming implementation use any unsigned integer sufficient to store the values 0-31?

woodard commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,260 @@
+>   referred to as the `overlay base`.  It begins at `base location`
+>   offset by `base offset` and has a size of `overlay width`.
+>
+>   *If the `overlay width` is zero. Then the consumer may leave the
+>   `base location` on the top of the stack rather than creating
+>   composite storage.*
```
I added:  If the `overlay width` is zero and offset is within the bounds of  the base location's storage, then...

woodard committed:
Apply fixes suggesed by Tony.

Tony spotted several errors and had a few comments. I applied almost
all of his suggested fixes and replied to the others but left them
unresolved awaiting further discussion.

Signed-off-by: Ben Woodard <woodard@redhat.com>

ccoutant commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,261 @@
+
+**NOTE:** I'm not really buying the non-normative explanation. I think it
+was a mistake and you should flat out say that and possibly suggest
+`DW_OP_const_type2` which drops the size operand.
```
No, a DWARF consumer truly cannot parse the expression without knowing the size of the constant. Without that second parameter, the consumer would be forced to go parse the `.debug_info` section to extract the size before it can continue parsing the expression.

The mistake was copying this size parameter, and the same non-normative text, over to the `DW_OP_deref_type` operator, where it really is unnecessary. We've removed the text from the `DW_OP_deref_type` operator in the locations-on-stack proposal, but we haven't removed the redundant second operand. That's where you should press for admitting our mistake and fixing it! The opcode space is scarce enough, though, that I don't think it's worth burning an opcode just to eliminate the one-byte size parameter.

woodard replied:
Not to be too argumentative but don't most if not all consumers need to have a hard coded understanding of the base types to enable an architecture otherwise they would not know what the encoding of value that follows. 

Consider the absurdity of something like having IEEE754 decimal floating point number as a value. It is a base type on some processors. As far as I know there is no place in DWARF where the encoding of values is specified even though the size of base types is specified in the CU's DWARF, thus the encoding for all architectures base types must be included in hard coded data about the architecture. 

Since this operation is limited to the base types, will they actually have to refer to the .debug_info to know the type's size or do they already know it? 

Either way, your point is made. You don't want to burn an opcode just to eliminate a 1-byte parameter. This is not a hill that I want to die on. I'll drop the text.

ccoutant replied:
Yes, it must be a base type, but you don't know what base type it's referring to without veering off to look at the `.debug_info` section. The DIE offset by itself doesn't tell you — you have to go read the DIE at that offset. All that just to know how many bytes to consume to get to the next operator.

palves commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,260 @@
+>      - stack parameters:
+>        - T [location]: base location
+>        - Z [location]: overlay location
+>        - Y [unsigned int]: base offset
```
"base offset" is a relative offset, very much like the offset passed to DW_OP_offset is a relative offset.  DW_OP_offset takes a signed offset.  Making this one be signed too is more consistent than not.  I don't think CVEs are relevant here.  You can go out of bounds with an unsigned offset too.  ptrdiff_t exists for a reason.

woodard committed:
Continue formatting operators

Continue the process of converting operators to the new format.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Finish updating operators for 3.7

Finish updating the operators through section 3.7.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Update operators through 3.10

Continue updating operators to the new format up through section 3.10.
Also fix a couple of mistakes made in earlier sections.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Update operations through section 3.12

Update the operations descriptions to the new format up through
section 3.12 the composite operators.
Delete a few trailing whitespaces that have crept in.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Update descriptions through 3.15

Update the descriptions to the new format up through section 3.15.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Finish first draft of operator reformatting

- Per Cary's request change the terminology to "inline parameters" and
"stack operands". This may end up being too big of a change to make
outside of chapter 3 of the standard.
- Finish  off the rest of the operators.
- Remove discussion comments for issues that are settled.
- Try to make everything consistent.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
fix a paramter/operand swap.

Fix a place where I missed converting from parameter to operand.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge branch 'upstream' into op-formatting

woodard committed:
Add table 8.9 to changes.

Table 8.9 currently has an "operand" column. With the new terminology
this should now be called "parameters".

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Try out a new header format.

This reverts the changes of the name used for inline operands and
parameters. This was just too big of a change to make to the DWARF
standard and would have impacted a broad range of chapters not just
chapter 3. Without that change, the descriptions of the the operands
doesn't need to change and so the change ends up being just add a
header to each header. This makes this change much smaller.

With a header like this, many descriptions of the operators can be
greatly simplified. The sources of the information used by the
operators, their types as well as the output of the operator can be
removed from the operator description leaving the description to
describe the function of the operator.

Try out a new header format. Rather than a "bulky" set of lists inside
of list this compresses all of the information into a line.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Change operation header to version 3

Ron and Cary suggested some changes to my header format.
1) Get rid of the numbers
2) Put the inline operands on the operation declaration line
3) Remove the operation name from the stack effects line.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Finish making requested changes for V3

Finish, making the requested changes for v3 of this draft.

* Added additional expository text to the beginning of chapter 3. This
  helps clarify the difference between inline operands where type
  refers to their encoding and stack arguments where type has more to
  do with limitations on operator's domain.
* Change to using A B C D vs. X Y Z T
* Added a few more discussion points as I considered them
* Switched from LEB128 and ULEB128 to SLEB and ULEB, This change will
  need to be propegated throughout the rest of the DWARF standards
  document.
* Removed errant prime on uconst operators.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Further refinements. Let's call this v3.1

Ron pointed out some problems. I tried to address them.
1) I had a spurious 2 in one place. I blame the cat.
2) I had given dup a spurious stack argument.
3) Added some text to explain that the stack grows to the right and
thus the top of the stack is on the right while the deepest part of
the stack is on the left. I also went through and made sure that I
didn't reverse the stack in any operators.
4) By convention I made A the deepest stack argument. I had not done
this consistently.

Signed-off-by: Ben Woodard <woodard@redhat.com>

markus-metzger commented on 024-revise-operations.md:
```diff
@@ -0,0 +1,521 @@
+
+> `DW_OP_over`
+>
+>     <[*any*] A> <[*any*] B> → < B> < A> < B>
```
Shouldn't the result be `<A> <B> <A>`?  L55-61 explains that the top of stack is on the right.

woodard committed:
Address problmes that Markus pointed out.

Markus Metzger found 3 problems which pointed out another
problem. This patch addresses the issues that he pointed out.

When converting to the new convention where A is always on the left I
introduced a couple of bugs.
1) I had described drop incorrectly.
2) I had swapped arguments in the output of over.

I also had failed to paste in some boiler plate that I was using for
the desctiption of the changes to be made to section 3.2

His suggested change to swap also pointed out the need to explain that
generic stack argument names referred to the data and its type rather
than giving the stack position a name.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Some additional tweaks

These are just some minor tweaks to the header proposal.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge branch 'upstream' into op-formatting

woodard committed:
Merge branch 'main' into op-formatting

woodard committed:
Final revision

Final revision of the DWARF Operation Header's proposal.
  * Remove FIXME's - moved those to 005a
  * Update README.md

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge branch 'upstream' into op-formatting

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge branch 'upstream' into op-formatting

woodard committed:
Fix omission on DW_OP_bit_piece

Pedro noticed that I failed to update bit piece the way that I had
updated piece. Fixed.

Signed-off-by: Ben Woodard <woodard@redhat.com>

woodard committed:
Merge branch 'upstream' into op-formatting

woodard committed:
Add update pointing to Cary's graphical diagrams

Cary Coutant provided a TeX macro which draws standardized op diagrams
and demonstrated how to use it by having it draw the diagrams for all
the current operators.

This updates the issue to point to these files so that it can easily
be edited into the DWARF standard draft.

Signed-off-by: Ben Woodard <woodard@redhat.com>
